### PR TITLE
Fix for NPE when parsing

### PR DESCRIPTION
**Pull request by PaulHatcherR3**
CORDA-3756 fix for NPE on startup. Added test to expose issue and fix.

### CODE REVIEW

File: quasar-core/src/test/java/co/paralleluniverse/fibers/instrument/MethodDatabaseDuplicate.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2008-2013, Matthias Mann
```
Comment (chrisr3) [inline quasar-core/src/test/java/co/paralleluniverse/fibers/instrument/MethodDatabaseDuplicate.java:2]: Who is this guy? Did he write this test?

File: build.gradle [hunk 2]
```diff
@@ -422,7 +422,7 @@ project (':quasar-core') {
         jdk8test {
             java {
                 srcDir 'src/test/java'
-                srcDir 'src/jdk8Test/java'
+                srcDir 'src/jdk8test/java'
```
Comment (chrisr3) [inline build.gradle:425]: Well spotted. This must mean that we haven't been running the JDK8 tests recently...

File: quasar-core/src/test/java/co/paralleluniverse/fibers/instrument/MethodDatabaseDuplicate.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+ * This leads to duplicate class name in the database but with different ClassEntry objects.
+ * When testing ClassEntry objects for equality the equals operator is blowing up as the superName can be null.
+ * @author Paul Hatcher
+ */
+@Instrumented
+public class MethodDatabaseDuplicate {
+
+    @Test
+    public void testSuspend() throws IOException {
+        final String className = MethodDatabaseDuplicate.class.getName().replace('.', '/');
```
Comment (chrisr3) [inline quasar-core/src/test/java/co/paralleluniverse/fibers/instrument/MethodDatabaseDuplicate.java:53]:
```java
import org.objectweb.asm.Type;

final String className = Type.getInternalName(MethodDatabaseDuplicate.class);
```

File: quasar-core/src/test/java/co/paralleluniverse/fibers/instrument/MethodDatabaseDuplicate.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+package co.paralleluniverse.fibers.instrument;
+
+import co.paralleluniverse.fibers.Instrumented;
+import co.paralleluniverse.fibers.SuspendExecution;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashSet;
+import java.util.Locale;
+import org.junit.Test;
+import static org.junit.Assert.*;
```
Comment (chrisr3) [inline quasar-core/src/test/java/co/paralleluniverse/fibers/instrument/MethodDatabaseDuplicate.java:38]: Can you remove the unused imports please?

File: quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/MethodDatabase.java [hunk 5]
```diff
@@ -554,7 +555,8 @@ public boolean equals(Object obj) {
                 return false;
             }
             final ClassEntry other = (ClassEntry) obj;
-            return superName.equals(other.superName) && methods.equals(other.methods);
+            // CORDA-3756 names can be null.
+            return Objects.equals(superName, other.superName) && methods.equals(other.methods);
```
Comment (chrisr3) [inline quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/MethodDatabase.java:559]: OK, so this resolves the NPE. What does a `MethodDatabase` entry without a `superName` mean for Quasar though?

**Comment by chrisr3** [comment]
No NPEs _must_ be a Good Thing :tm:. But should a database entry without a `superName` exist in the first place?

**Comment by chrisr3** [comment]
According to the ASM JavaDoc:
```java
/**
 * @param superName the internal of name of the super class (see {@link Type#getInternalName()}).
 *     For interfaces, the super class is {@link Object}. May be {@literal null}, but only for the
 *     {@link Object} class.
 */
```
I'd misremembered this - apparently only `java.lang.Object` should have a `null` superclass.

File: quasar-core/src/test/java/co/paralleluniverse/fibers/instrument/MethodDatabaseDuplicate.java (hunk 1 above)
Comment (PaulHatcherR3) [reply quasar-core/src/test/java/co/paralleluniverse/fibers/instrument/MethodDatabaseDuplicate.java:2]: Yes, spotted the comment post commit will remove, one of the delights of folding editors ;)

File: build.gradle (hunk 2 above)
Comment (PaulHatcherR3) [reply build.gradle:425]: Thanks! On an OSX platform would not build correctly, took me a while to figure out what was wrong.

**Comment by PaulHatcherR3** [comment]
"Does the NPE only happen when comparing a ClassEntry for java.lang.Object?" Yes I believe that is the case, as superName is null. If the program being instrumented is single threaded it cannot occur because in MethodDatabase.recordSuspendableMethods there cannot be another member of classes with a different ClassEntry object for the given className.

However, in a multithreaded process there is a race condition in MethodDatabase where two threads can enter different ClassEntry objects for the same className. Now normally this results in the warning message "Duplicate class entries with different data for class". However, if you're unlucky and className is "java.langObject" then ClassName.SuperName is null and the NPE occurs in ClassEntry.equals on comparison.

So this is why we see this problem more in multi core hosts with finer grain parallelism.

**Comment by chrisr3** [comment]
> "Does the NPE only happen when comparing a ClassEntry for java.lang.Object?" Yes I believe that is the case, as superName is null.

I guess I'm asking whether you have _confirmed_ that the `ClassEntry` is always for `java.lang.Object` whenever you reproduce the problem.

**Comment by PaulHatcherR3** [comment]
Just to confirm yes, I've added a bit more logging in the class and can see this only occurs for me with java.lang.Object s.

**Pull request merged**